### DISCUSSION TITLE
parallellization choice

### DISCUSSION THREAD

**Discussion by chenqin** [varchar-io/nebula#66] [Ideas]
Nebula has imported [google highway](https://github.com/google/highway) to utilize SIMD in capability across various of instruction sets. The pro of this approach is small, portability and controllable behavior than -03 compiler setting. It appears the library mostly used to accelerate matrix manipulation and image encode/decode related tasks for set of google internal projects.

In another route, in xgboost project, we have been using openmp to accelerate subroutines with good result (multi-threading) and for loops. OpenMP is well established standard and well supported by almost all vendors in industry. Adopting openmp could help us leverge async/threading/simd/synchornization issues and focus on building core features runs everywhere.

https://ecpannualmeeting.com/assets/overview/sessions/ff2020%20ECP-Tutorial-with-ECP-template.pdf
https://stackoverflow.com/questions/14674049/parallel-for-vs-omp-simd-when-to-use-each

**Comment by shawncao** [comment]
OpenMP looks great. I think first thing is to evolve current value-based processing interface into array-based processing and then we can leverage those tools to make it from fast to extremely fast, :)

**Reply by shawncao** [reply]
@chenqin Could you talk about this a bit more when we meet next time? very interesting topic and I don't know much about it.
